# **1. Data Ingestion**

**1.1 Installations and Imports**

In [16]:
# Necesssary installation for scikeras.wrappers.KerasClassifier
!pip install scikeras

In [48]:
import keras
from keras.datasets import mnist
from keras.layers import Conv2D, MaxPooling2D
from keras.layers import Dense, Flatten
from keras.models import Sequential
from sklearn.pipeline import Pipeline
from keras import layers, models
from keras.optimizers import Adam
from scikeras.wrappers import KerasClassifier
from sklearn.model_selection import GridSearchCV
from sklearn.metrics import accuracy_score

**1.2 Loading the Training adn Out-of-sample Data**

In [2]:
(x_train, y_train), (x_test, y_test) = mnist.load_data()

11490434/11490434 [==============================] - 0s 0us/step


# **2. Data Preprocessing**

2.1 One Hot Encoding for categorical data

In [3]:
# No. of classes = max(y) + 1
y_train = keras.utils.to_categorical(y_train)
y_test = keras.utils.to_categorical(y_test)

# **3. Defining the CNN Model**

**3.1 LeNet Model Architecture**

In [39]:
def build_model(filters, learning_rate):
  model = models.Sequential()
  #define filters and convolutional layers here
  model.add(layers.Conv2D(filters=filters, kernel_size=(3, 3), activation='relu',
  input_shape=(28, 28, 1)))
  #Add a maxpooling layer
  model.add(layers.MaxPooling2D(pool_size=(2, 2)))
  #Flatten the output and give it to a fully connected layer
  model.add(layers.Flatten())
  #one hidden layer maps the flattened neurons to output
  model.add(layers.Dense(10, activation='softmax'))
  optimizer = Adam(learning_rate=learning_rate)
  model.compile(optimizer=optimizer, loss='categorical_crossentropy',
  metrics=['accuracy'])
  return model

# **4. Parameter Exploration**

**4.1 Using GridSearchCV for stratified cross validation**

In [43]:
# Using Keras Classifier API to make our Keras model compatible with GridSearchCV
model = KerasClassifier(build_fn=build_model,
                        filters=16,
                        learning_rate=0.001,
                        verbose=1)
pipeline = Pipeline(steps=[('model', model)])
learning_rate=[0.001, 0.01]
filters = [16, 32]
param_grid = dict(filters=filters, learning_rate=learning_rate)
grid = GridSearchCV(estimator=pipeline.named_steps['model'], param_grid=param_grid, cv=5)
grid_result = grid.fit(x_train, y_train)

/usr/local/lib/python3.10/dist-packages/scikeras/wrappers.py:915: UserWarning: ``build_fn`` will be renamed to ``model`` in a future release, at which point use of ``build_fn`` will raise an Error instead.
  X, y = self._initialize(X, y)


375/375 [==============================] - 1s 4ms/step


/usr/local/lib/python3.10/dist-packages/scikeras/wrappers.py:915: UserWarning: ``build_fn`` will be renamed to ``model`` in a future release, at which point use of ``build_fn`` will raise an Error instead.
  X, y = self._initialize(X, y)


375/375 [==============================] - 1s 4ms/step


/usr/local/lib/python3.10/dist-packages/scikeras/wrappers.py:915: UserWarning: ``build_fn`` will be renamed to ``model`` in a future release, at which point use of ``build_fn`` will raise an Error instead.
  X, y = self._initialize(X, y)


375/375 [==============================] - 2s 6ms/step


/usr/local/lib/python3.10/dist-packages/scikeras/wrappers.py:915: UserWarning: ``build_fn`` will be renamed to ``model`` in a future release, at which point use of ``build_fn`` will raise an Error instead.
  X, y = self._initialize(X, y)


375/375 [==============================] - 2s 4ms/step


/usr/local/lib/python3.10/dist-packages/scikeras/wrappers.py:915: UserWarning: ``build_fn`` will be renamed to ``model`` in a future release, at which point use of ``build_fn`` will raise an Error instead.
  X, y = self._initialize(X, y)


375/375 [==============================] - 1s 4ms/step


/usr/local/lib/python3.10/dist-packages/scikeras/wrappers.py:915: UserWarning: ``build_fn`` will be renamed to ``model`` in a future release, at which point use of ``build_fn`` will raise an Error instead.
  X, y = self._initialize(X, y)


375/375 [==============================] - 1s 4ms/step


/usr/local/lib/python3.10/dist-packages/scikeras/wrappers.py:915: UserWarning: ``build_fn`` will be renamed to ``model`` in a future release, at which point use of ``build_fn`` will raise an Error instead.
  X, y = self._initialize(X, y)


375/375 [==============================] - 1s 4ms/step


/usr/local/lib/python3.10/dist-packages/scikeras/wrappers.py:915: UserWarning: ``build_fn`` will be renamed to ``model`` in a future release, at which point use of ``build_fn`` will raise an Error instead.
  X, y = self._initialize(X, y)


375/375 [==============================] - 1s 4ms/step


/usr/local/lib/python3.10/dist-packages/scikeras/wrappers.py:915: UserWarning: ``build_fn`` will be renamed to ``model`` in a future release, at which point use of ``build_fn`` will raise an Error instead.
  X, y = self._initialize(X, y)


375/375 [==============================] - 2s 6ms/step


/usr/local/lib/python3.10/dist-packages/scikeras/wrappers.py:915: UserWarning: ``build_fn`` will be renamed to ``model`` in a future release, at which point use of ``build_fn`` will raise an Error instead.
  X, y = self._initialize(X, y)


375/375 [==============================] - 1s 4ms/step


/usr/local/lib/python3.10/dist-packages/scikeras/wrappers.py:915: UserWarning: ``build_fn`` will be renamed to ``model`` in a future release, at which point use of ``build_fn`` will raise an Error instead.
  X, y = self._initialize(X, y)


375/375 [==============================] - 2s 4ms/step


/usr/local/lib/python3.10/dist-packages/scikeras/wrappers.py:915: UserWarning: ``build_fn`` will be renamed to ``model`` in a future release, at which point use of ``build_fn`` will raise an Error instead.
  X, y = self._initialize(X, y)


375/375 [==============================] - 3s 7ms/step


/usr/local/lib/python3.10/dist-packages/scikeras/wrappers.py:915: UserWarning: ``build_fn`` will be renamed to ``model`` in a future release, at which point use of ``build_fn`` will raise an Error instead.
  X, y = self._initialize(X, y)


375/375 [==============================] - 3s 7ms/step


/usr/local/lib/python3.10/dist-packages/scikeras/wrappers.py:915: UserWarning: ``build_fn`` will be renamed to ``model`` in a future release, at which point use of ``build_fn`` will raise an Error instead.
  X, y = self._initialize(X, y)


375/375 [==============================] - 2s 6ms/step


/usr/local/lib/python3.10/dist-packages/scikeras/wrappers.py:915: UserWarning: ``build_fn`` will be renamed to ``model`` in a future release, at which point use of ``build_fn`` will raise an Error instead.
  X, y = self._initialize(X, y)


375/375 [==============================] - 2s 4ms/step


/usr/local/lib/python3.10/dist-packages/scikeras/wrappers.py:915: UserWarning: ``build_fn`` will be renamed to ``model`` in a future release, at which point use of ``build_fn`` will raise an Error instead.
  X, y = self._initialize(X, y)


375/375 [==============================] - 2s 4ms/step


/usr/local/lib/python3.10/dist-packages/scikeras/wrappers.py:915: UserWarning: ``build_fn`` will be renamed to ``model`` in a future release, at which point use of ``build_fn`` will raise an Error instead.
  X, y = self._initialize(X, y)


375/375 [==============================] - 2s 4ms/step


/usr/local/lib/python3.10/dist-packages/scikeras/wrappers.py:915: UserWarning: ``build_fn`` will be renamed to ``model`` in a future release, at which point use of ``build_fn`` will raise an Error instead.
  X, y = self._initialize(X, y)


375/375 [==============================] - 2s 4ms/step


/usr/local/lib/python3.10/dist-packages/scikeras/wrappers.py:915: UserWarning: ``build_fn`` will be renamed to ``model`` in a future release, at which point use of ``build_fn`` will raise an Error instead.
  X, y = self._initialize(X, y)


375/375 [==============================] - 3s 7ms/step


/usr/local/lib/python3.10/dist-packages/scikeras/wrappers.py:915: UserWarning: ``build_fn`` will be renamed to ``model`` in a future release, at which point use of ``build_fn`` will raise an Error instead.
  X, y = self._initialize(X, y)


375/375 [==============================] - 2s 6ms/step


/usr/local/lib/python3.10/dist-packages/scikeras/wrappers.py:915: UserWarning: ``build_fn`` will be renamed to ``model`` in a future release, at which point use of ``build_fn`` will raise an Error instead.
  X, y = self._initialize(X, y)


1875/1875 [==============================] - 24s 13ms/step - loss: 0.5333 - accuracy: 0.9399


**4.2 Best accuracy and parameters**

In [44]:
[best_score, best_params] = grid_result.best_score_, grid_result.best_params_
print([best_score, best_params])

[0.9622999999999999, {'filters': 32, 'learning_rate': 0.001}]


When our CNN model has 32 filters and 0.001 learning rate, our model's training accuracy is 96.23%

# **5. Training the best model**

Training our CNN model using the best parameters on out-of-sample data

In [53]:
best_model=KerasClassifier(build_fn=build_model,
                        filters=32,
                        learning_rate=0.001,
                        verbose=1)
pipeline = Pipeline(steps=[('model', best_model)])
pipeline.fit(x_train, y_train)

/usr/local/lib/python3.10/dist-packages/scikeras/wrappers.py:915: UserWarning: ``build_fn`` will be renamed to ``model`` in a future release, at which point use of ``build_fn`` will raise an Error instead.
  X, y = self._initialize(X, y)


1875/1875 [==============================] - 24s 13ms/step - loss: 0.5348 - accuracy: 0.9426


Pipeline(steps=[('model',
                 KerasClassifier(build_fn=<function build_model at 0x7c5d74374430>, filters=32, learning_rate=0.001))])

# **6. Evaluating the best model's performance**

Using out-of-sample data set to for making predictions and evaluating the best model's performance on it.
- x_test : out-of-sample values used for prediction
- y_predicted : predictions made on x_test
- accuracy_score : accuracy of the best model is measured based on how accurate y_predicted is, given y_test are the true values

In [54]:
y_predicted = pipeline.predict(x_test)
accuracy = accuracy_score(y_test, y_predicted)

313/313 [==============================] - 1s 4ms/step


In [55]:
print(accuracy)

0.9685


The validation accuracy (accuracy on out-of-sample data) of our model using the best parameters (32 filters and 0.001 learning rate) is 96.85%